cobalt/watchdog/instrumentation_log.cc

```diff
@@ -0,0 +1,65 @@
+// Copyright 2024 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cobalt/watchdog/instrumentation_log.h"
+
+#include <string>
+#include <vector>
+
+namespace cobalt {
+namespace watchdog {
+
+bool InstrumentationLog::LogEvent(const std::string& event) {
+  if (event.length() > kMaxEventLen) {
+    SB_DLOG(ERROR) << "[Watchdog] Log event exceeds max: " << kMaxEventLen;
+    return false;
+  }
+
+  starboard::ScopedLock scoped_lock(buffer_mutex_);
+  if (buffer_.CurrentIndex() == 0 || event != **(buffer_.End())) {
```

Comment:
I guess my question with this is, what if someone wants to add the same event to the buffer multiple times. For example imagine a loop or something like that. Or just the same function called twice.

Reply:
my intention is to avoid exactly this. But I see the point: some user may not want this optimization and want literally a snapshot of what was passed to the `logEvent()`. I think I can implement this de-duping on client side instead by caching the previous event and comparing. And this would be safer as well.